snow/networking/blacklist/blacklist.go

```diff
@@ -0,0 +1,231 @@
+package blacklist
+
+import (
+	"container/list"
+	"sync"
+	"time"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/validators"
+	"github.com/ava-labs/avalanchego/utils/timer"
+
+	safemath "github.com/ava-labs/avalanchego/utils/math"
+)
+
+// If a peer consistently does not respond to queries, it will
+// increase latencies on the network whenever that peer is polled.
+// If we cannot terminate the poll early, then the poll will wait
+// the full timeout before finalizing the poll and making progress.
+// This can increase network latencies to an undesirable level.
+
+// Therefore, a blacklist is used as a heurstic to immediately fail
+// queries to nodes that are consistently not responding.
+
+type queryBlacklist struct {
+	vdrs validators.Set
+	// Map of peerIDs to pending query requests
+	pendingQueries map[[20]byte]map[uint32]struct{}
+	// Map of consecutive query failures
+	consecutiveFailures map[[20]byte]int
+
+	// Maintain blacklist
+	blacklistTimes map[[20]byte]time.Time
+	blacklistOrder *list.List
+	blacklistSet   ids.ShortSet
+
+	threshold  int
+	duration   time.Duration
+	maxPortion float64
+
+	clock timer.Clock
+
+	lock sync.Mutex
+}
+
+// Config defines the configuration for a blacklist
+type Config struct {
+	Validators validators.Manager
+	Threshold  int
+	Duration   time.Duration
+	MaxPortion float64
+}
+
+// NewQueryBlacklist ...
+func NewQueryBlacklist(validators validators.Set, threshold int, duration time.Duration, maxPortion float64) QueryBlacklist {
+	return &queryBlacklist{
+		pendingQueries:      make(map[[20]byte]map[uint32]struct{}),
```

Comment:
` // Validator ID --> Request ID --> non-nil iff there is an outstanding request to this validator. `

Reply:
Added